File: 106. Construct Binary Tree from Inorder and Postorder Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def buildTree(self, inorder, postorder):
        """
        :type inorder: List[int]
        :type postorder: List[int]
        :rtype: TreeNode
        """
        if len(inorder) == 0:
            return None
        ind = inorder.index(postorder[len(postorder)-1])
        postorder.pop()
        root = TreeNode(inorder[ind])
        left = inorder[:ind]
        right = inorder[ind+1:]
        root.right = self.buildTree(right, postorder)
        root.left = self.buildTree(left, postorder)
        return root

File: 106. Construct Binary Tree from Inorder and Postorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        if (inorder.size() == 0) return NULL;
        int ind = find(inorder.begin(), inorder.end(), postorder[postorder.size()-1]) - inorder.begin();
        postorder.erase(postorder.end()-1);
        TreeNode* root = new TreeNode(inorder[ind]);
        vector <int> left(inorder.begin(), inorder.begin()+ind);
        vector <int> right(inorder.begin()+ind+1, inorder.end());
        root->right = buildTree(right, postorder);
        root->left = buildTree(left, postorder);
        return root;
    }
};
